new/resource-actions.md

```diff
@@ -0,0 +1,199 @@
+run.
+
+- The entire action recipe is compiled (just like a normal recipe) before
+  converging.
+- This compile happens when the action is *run*, not when the parent resource
+  is declared.
+- Delayed and immediate notifications are local to the action.
+
+This is equivalent to `use_inline_providers` in a Provider, and in fact is
+implemented that way.  We just call out the behavior here to make it clear what will happen.
+
+#### Action composition and inheritance
+
+The action recipes are all added to a `Provider` class which inherits from the
+parent resource's provider class.  You may call `super` to call the parent
+class's version of the current action.
+
+```ruby
+class MyFile < Chef::Resource::File
+  action :create do
+    puts "I am a file yo"
+    super
+  end
+end
+```
+
+### Inline resource output
+
+When `use_inline_resources` runs, the output can be much more verbose than the
+user expects.  Particularly, they see a number of resources and actions that they didn't declare:
+
+```
+* x[blah] action blah
+  * file[/Users/jkeiser/x.txt] action create (up to date)
+   (up to date)
+```
+
+We suggest not actually showing sub-resources as separate things in output:
+
+```
+* x[blah] action blah (up to date)
+```
+
+An "update" would look like this (again, removes the extra nesting):
+
+```
+* x[blah] action blah
+  - create new file /Users/jkeiser/x.txt
+  - update content in file /Users/jkeiser/x.txt from none to a591a6
+  --- /Users/jkeiser/x.txt	2015-04-29 08:15:14.000000000 -0700
+  +++ /Users/jkeiser/.x.txt20150429-56731-6xk0lx	2015-04-29 08:15:14.000000000 -0700
+  @@ -1 +1,2 @@
+  +Hello World
+```
+
+The principle is: the user doesn't see resources they don't type.
+
+### inline_recipe resource
+
+People need to support test-and-set and why-run in their resources. Given that
+actions are defined as recipes, the typical (and best) way to do that is
+resources (which have `not_if` and `only_if`, etc.). These resources work well,
+but sometimes you want to do a clump of actions based on a single `if`, or you
+want to nest them.
+
+`inline_recipe` is designed to do that job.  resource lets you write recipe
+code. It also has the parent in scope (so you can still access instance
+variables). The resources inside will compile immediately and be available in
+`inline_recipe.resources`.  The resources will converge when the `inline_recipe`
+converges.
+
+Like recipe actions, the resources inside the `inline_recipe` are not
+referenceable by `notifies` outside.  Unlike recipe actions, the resources show
+up in nested output (because the user typed them in the recipe).
+
+```ruby
+inline_recipe 'Create that one file' do
+  only_if { File.exist?('/please_create_that_file.txt') }
+
+  directory '/x' do
+  end
+  file '/x/y.txt' do
+  end
+end
+```
+
+`converge_by` (the current way of converging / updating) a resource will still
+be available for those who are used to it.
```

Comment:
I'm not certain I find the use case totally compelling.  And it smells like something which you would probably want to do within the current resource collection and not open up a new resource collection context.   If you're not doing notifies/subscribes to those resources then you could just use an if block to make the resources go away.

It also feels like a use case that would be better handled by wrapping resources with a block of defaults like you wanted to do with user/group/mode to prevent repetitive typing.